src/tennis/core.clj

```diff
@@ -0,0 +1,131 @@
+(ns tennis.core)
+
+(def config {:max-matches 3
+             :labels [0 15 30 40 "A"]})
+
+
+(def default-scores {:player1 [0 [0] 0]
+                     :player2 [0 [0] 0]})
+
+(def scores (atom default-scores))
+
+
+(defn inc-game-only?
+  "when winning a game increases only the game score"
+  [game game-adv]
+  (or (< game 3)
+      (= game game-adv)))
+
+
+(defn inc-set-match?
+  "when winning a game leads to increase set and match scores"
+  [game game-adv set set-adv]
+  (and (>= (- game game-adv) 1)
+       (>= set 5)
+       (>= (- set set-adv) 1)))
+
+
+(defn inc-set-only?
+  "when winning a game increases only set score"
+  [game game-adv set set-adv]
+  (and (>= game 3)
+       (>= (- game game-adv) 1)
+       (or (< set 6)
+           (= (- set set-adv) -1)
+           (= (- set set-adv) 0))))
+
+
+(defn adversary-advantage?
+  "when adversary has advantage, then it defaults to 40-40(deuce) for both players"
+  [game game-adv]
+  (= (- game game-adv) -1))
+
+
+(defn increase-score
+  ([player] (increase-score player @scores))
+  ([player score]
+   (let [[game sets match] (player score)
+         [adversary [game-adv sets-adv match-adv]] (->> (remove #(= (key %) player) score)
+                                                        (first))
+         set (last sets)
+         set-adv (last sets-adv)
+         {:keys [max-matches]} config]
+     (cond
+       (>= (max match match-adv) max-matches)
+       score
+
+         ;;when winning a game increases only the game score 
+       (inc-game-only? game game-adv)
```

Comment:
If possible, to make it a bit more readable I would change the order of conditions here:
```
adversary-advantage?
inc-game-only?
inc-set-only?
inc-set-match?
```